### DISCUSSION TITLE
multi-line def statement header 들여쓰기 정책

### DISCUSSION THREAD

**Discussion by cbscsm** [8percent/styleguide#39] [General]
저희 코드에서 여러 줄에 걸친 def문의 인수 들여쓰기에 대해서, 네 가지 방법(아래 1~4번)이 혼용되고 있습니다.
1. 여는 괄호에 맞춰 들여쓰기 + 마지막 인자와 닫는 괄호 함께 쓰기 -> PEP8 및 현재 컨벤션에 있음
```python
    def foovar(arg_1,
               arg_2,
               arg_3):
        ...
```
2. 여는 괄호에 맞춰 들여쓰기 + 닫는 괄호를 독립된 행에 표기
```python
    def foovar(arg_1,
               arg_2,
               arg_3,
               ):
        ...
```
3. 4칸 들여쓰기 + 닫는 괄호를 독립된 행에 표기(header와 들여쓰기 맞춤) -> black 스타일
```python
    def foovar(
        arg_1,
        arg_2,
        arg_3,
    ):
        ...
```
4. 8칸 들여쓰기 + 닫는 괄호를 독립된 행에 표기(header와 들여쓰기 맞춤)
```python
    def foovar(
            arg_1,
            arg_2,
            arg_3,
    ):
        ...
```
5. 8칸 들여쓰기 + 닫는 괄호를 독립된 행에 표기(suite와 들여쓰기 맞춤) -> 현재 컨벤션에 있음, 코드에는 없음
```python
    def foovar(
            arg_1,
            arg_2,
            arg_3,
        ):
        ...
```
6. 8칸 들여쓰기 + 마지막 인자와 닫는 괄호 함께 쓰기 -> PEP8에 있음, 코드에는 (아마도) 없음
```python
    def foovar(
            arg_1,
            arg_2,
            arg_3):
        ...
```
현재 저희 [내부 컨벤션](https://github.com/8percent/styleguide/tree/master/Python#multi-line-parameter)에서는 1번과 5번을 허용하고 있습니다.
그러나 저희 코드 중에서는 5번 케이스는 찾을 수 없었고, 1~4번 케이스가 모두 발견됩니다.
1번 [예시](https://github.com/8percent/eight/blob/master/apps/deals/tests/test_queryset.py#L90-L96)
2번 - 6건, [예시](https://github.com/8percent/eight/blob/6d02874c74235bfc498306f099488c9dcaf8a1bf/apps/payments/models/payback.py#L408-L414) <- 레거시로 판단됩니다(일부 trailing comma가 빠진 경우는 더더욱).
3번 [예시](https://github.com/8percent/eight/blob/6d02874c74235bfc498306f099488c9dcaf8a1bf/eight_staff/repayments/tests/test_views.py#L119-L127)
4번 [예시](https://github.com/8percent/eight/blob/6d02874c74235bfc498306f099488c9dcaf8a1bf/api/aml/tests/test_views.py#L467-L477)

- 이와 관련해서는 [커뮤니티 논의](https://discuss.python.org/t/pep-8-clarify-if-multiline-argument-list-with-a-closing-on-a-separate-line-is-acceptable/2948)가 있기도 했습니다.
- [PEP8](https://peps.python.org/pep-0008/#indentation)에서는 기본적으로 닫는괄호(`):`)를 독립된 행에 쓰는 것을 고려하고 있지 않으며, 때문에 4칸 들여쓰기(3번)를 금지하고 있습니다.
- PEP8은 1번과 6번만을 허용하지만 [파이썬 내부 개발자를 위한 (표준 라이브러리 작성) 가이드일 뿐](https://peps.python.org/pep-0008/#introduction) 외부 개발자들에게는 참고용이므로, 자체적으로 다른 규칙을 정해도 됩니다.
- black 에서는 3번 스타일을 사용합니다.
- 닫는 괄호를 독립된 행에 쓰면, 리턴 타입힌트와 trailing comma를 사용할 수 있다는 장점이 있습니다.
> [Black optimizes for a different goal than PEP 8; PEP 8 optimizes for readability while Black sometimes compromises readability in order to satisfy other goals such as minimizing lines changed as code evolves.](https://discuss.python.org/t/pep-8-clarify-if-multiline-argument-list-with-a-closing-on-a-separate-line-is-acceptable/2948/7)
- 닫는 괄호를 마지막 인자와 함께 쓰면, 닫는괄호가 어정쩡하게 홀로 있지 않아도 된다는 장점이 있습니다.

새로 작성되는 코드에서도 계속 다른 스타일이 사용되고 있고, 현재 컨벤션이 지켜지지 않고 있기 때문에,
컨벤션 재정의를 위해 논의를 해보면 좋을 것 같습니다.

**Comment by cbscsm** [comment]
- 앞으로 black을 도입할 여지가 있고 -> 8칸 보단 4칸 들여쓰기 + `):`를 독립된 행에 작성
- trailing comma 사용이 가능하며 -> `):`를 독립된 행에 작성
- (특히 긴) 리턴 타입 힌트 작성이 용이하다는 점에서 -> `):`를 독립된 행에 작성
- 기타 `list`, `set`, `dict` 등의 display 문법과도 비슷한 스타일이라는 점에서

(파이썬 내부 개발자들은 싫어하지만) 저는 3번도 좋을 것 같습니다.
위에서 언급한 기타 display문법들에서 허용되듯이, 1번 스타일 역시 좋을 것 같습니다.
개인적으로는 PEP8과 저희 컨벤션을 모두 만족하는 1번을 사용해왔습니다.

**Reply by Kirade** [reply]
저도 1, 3번을 선호합니다.

**Reply by wer153** [reply]
제시해주신 근거 모두에 동의하여 3번을 선호합니다.

**Reply by earthkingman** [reply]
저는 3번을 선호합니다.

**Reply by cbscsm** [reply]
생각해보니 가장 비슷하게 생긴 함수 호출의 경우도 1, 3번을 사용하네요.
```python
some_function(first,
              second,
              third)  # 1번

some_function(
    first,
    second,
    third,
)  # 3번
```
함수 호출에서도 1, 3번이 둘 다 허용 되어 상황에 맞게 쓸 수 있듯이, 함수 정의에서도 둘 다 허용하는 방법도 좋을 것 같습니다.

**Reply by twins1040** [reply]
Django 에 한정해서 보면 이미 최신 버전에서 black 이 적용되었기 때문에, 우리 코드도 black 을 도입하는 것이 좋을 것 같습니다.
적절한 시기에 도입하면 적어도 Django 에 대해서는 컨벤션 논의를 하지 않아도 될 것 같습니다.
